File: sdk/scala/src/main/scala/io/substrait/compliance/benchmark/BenchmarkExample.scala
package io.substrait.compliance.benchmark

import io.substrait.compliance._
import io.substrait.compliance.examples.ExampleEngine
import scala.concurrent.{Future, Await, ExecutionContext}
import scala.concurrent.duration._
import java.nio.file.{Files, Paths}

/**
 * Example demonstrating performance benchmarking
 */
object BenchmarkExample {
  import scala.concurrent.ExecutionContext.Implicits.global

  def main(args: Array[String]): Unit = {
    println("=== Substrait Compliance Benchmarking Example ===\n")

    // Create engine instance
    val engine = new ExampleEngine()

    // Run different benchmark scenarios
    runBasicBenchmark(engine)
    runPlanExecutionBenchmark(engine)
    runComparisonBenchmark()
    runScalabilityBenchmark(engine)

    println("\n=== Benchmarking Complete ===")
  }

  /**
   * Basic benchmark example
   */
  def runBasicBenchmark(engine: ComplianceEngine): Unit = {
    println("1. Basic Operation Benchmark")
    println("-" * 50)

    val config = BenchmarkConfig(
      warmupRuns = 10,
      measurementRuns = 100,
      verbose = true
    )

    val runner = BenchmarkRunner(engine, config)

    val operations = Seq(
      ("getInfo", () => Future.successful(engine.getInfo())),
      ("getCapabilities", () => Future.successful(engine.getCapabilities()))
    )

    val resultFuture = runner.runBenchmark("Basic Operations", operations)
    val result = Await.result(resultFuture, 1.minute)

    println(result.summary)
    println()
  }

  /**
   * Plan execution benchmark
   */
  def runPlanExecutionBenchmark(engine: ComplianceEngine): Unit = {
    println("2. Plan Execution Benchmark")
    println("-" * 50)

    // Create sample data
    val inputTable = TableData(
      columns = Seq(
        Column("id", ColumnType.Integer),
        Column("value", ColumnType.Float)
      ),
      rows = (1 to 1000).map(i => Seq(i, i * 1.5))
    )

    val plan = Array[Byte](1, 2, 3, 4, 5) // Mock plan

    val config = BenchmarkConfig(
      warmupRuns = 5,
      measurementRuns = 50,
      collectMemoryStats = true,
      verbose = true
    )

    val runner = BenchmarkRunner(engine, config)

    val operations = Seq(
      ("executePlan_1k_rows", () => 
        engine.executePlan(plan, Map("input" -> inputTable))),
      ("validatePlan", () => 
        engine.validatePlan(plan))
    )

    val resultFuture = runner.runBenchmark("Plan Execution", operations)
    val result = Await.result(resultFuture, 2.minutes)

    println(result.summary)
    
    // Export to CSV
    val csvPath = "benchmark_results.csv"
    Files.write(Paths.get(csvPath), result.toCSV.getBytes)
    println(s"Results exported to: $csvPath\n")
  }

  /**
   * Comparison benchmark between multiple engines
   */
  def runComparisonBenchmark(): Unit = {
    println("3. Engine Comparison Benchmark")
    println("-" * 50)

    // Create multiple engine instances
    val engines = Seq(
      new ExampleEngine(),
      new ExampleEngine(), // Could be different implementations
      new ExampleEngine()
    )

    val config = BenchmarkConfig(
      warmupRuns = 5,
      measurementRuns = 50,
      verbose = false
    )

    val runner = BenchmarkRunner(engines.head, config)

    val operations = Seq[(String, ComplianceEngine => Future[_])](
      ("getInfo", eng => Future.successful(eng.getInfo())),
      ("getCapabilities", eng => Future.successful(eng.getCapabilities()))
    )

    val resultsFuture = runner.compareBenchmark(
      "Engine Comparison",
      engines,
      operations
    )

    val results = Await.result(resultsFuture, 2.minutes)

    // Print comparison
    results.foreach { result =>
      println(s"\nEngine: ${result.engineName}")
      result.stats.foreach { stat =>
        println(f"  ${stat.operationName}: ${stat.avgTime.toMillis}ms avg, " +
                f"${stat.throughput}%.2f ops/sec")
      }
    }
    println()
  }

  /**
   * Scalability benchmark with varying data sizes
   */
  def runScalabilityBenchmark(engine: ComplianceEngine): Unit = {
    println("4. Scalability Benchmark")
    println("-" * 50)

    val dataSizes = Seq(100, 1000, 10000, 100000)
    val plan = Array[Byte](1, 2, 3, 4, 5)

    val config = BenchmarkConfig(
      warmupRuns = 3,
      measurementRuns = 20,
      verbose = false
    )

    println(f"${"Data Size"}%-15s ${"Avg Time"}%-15s ${"Throughput"}%-15s")
    println("-" * 50)

    dataSizes.foreach { size =>
      val inputTable = TableData(
        columns = Seq(
          Column("id", ColumnType.Integer),
          Column("value", ColumnType.Float)
        ),
        rows = (1 to size).map(i => Seq(i, i * 1.5))
      )

      val statsFuture = BenchmarkRunner.quickBenchmark(
        engine,
        s"executePlan_${size}_rows",
        () => engine.executePlan(plan, Map("input" -> inputTable)),
        runs = 20
      )

      val stats = Await.result(statsFuture, 1.minute)
      println(f"$size%-15d ${stats.avgTime.toMillis}%-15dms ${stats.throughput}%-15.2f ops/sec")
    }
    println()
  }

  /**
   * Parallel execution benchmark
   */
  def runParallelBenchmark(engine: ComplianceEngine): Unit = {
    println("5. Parallel Execution Benchmark")
    println("-" * 50)

    val parallelismLevels = Seq(1, 2, 4, 8)
    val plan = Array[Byte](1, 2, 3, 4, 5)

    val inputTable = TableData(
      columns = Seq(Column("id", ColumnType.Integer)),
      rows = (1 to 1000).map(i => Seq(i))
    )

    println(f"${"Parallelism"}%-15s ${"Avg Time"}%-15s ${"Throughput"}%-15s")
    println("-" * 50)

    parallelismLevels.foreach { parallelism =>
      val config = BenchmarkConfig(
        warmupRuns = 3,
        measurementRuns = 20,
        parallelism = parallelism,
        verbose = false
      )

      val runner = BenchmarkRunner(engine, config)

      val operations = Seq(
        ("executePlan", () => 
          engine.executePlan(plan, Map("input" -> inputTable)))
      )

      val resultFuture = runner.runBenchmark(
        s"Parallel_${parallelism}",
        operations
      )

      val result = Await.result(resultFuture, 2.minutes)
      val stats = result.stats.head

      println(f"$parallelism%-15d ${stats.avgTime.toMillis}%-15dms " +
              f"${stats.throughput}%-15.2f ops/sec")
    }
    println()
  }

  /**
   * Memory usage benchmark
   */
  def runMemoryBenchmark(engine: ComplianceEngine): Unit = {
    println("6. Memory Usage Benchmark")
    println("-" * 50)

    val config = BenchmarkConfig(
      warmupRuns = 5,
      measurementRuns = 50,
      collectMemoryStats = true,
      verbose = false
    )

    val runner = BenchmarkRunner(engine, config)

    val dataSizes = Seq(1000, 10000, 100000)

    dataSizes.foreach { size =>
      val inputTable = TableData(
        columns = Seq(
          Column("id", ColumnType.Integer),
          Column("data", ColumnType.String)
        ),
        rows = (1 to size).map(i => Seq(i, s"data_$i"))
      )

      val plan = Array[Byte](1, 2, 3, 4, 5)

      val statsFuture = runner.benchmarkOperation(
        s"memory_${size}_rows",
        () => engine.executePlan(plan, Map("input" -> inputTable))
      )

      val stats = Await.result(statsFuture, 2.minutes)
      println(f"Data size: $size%-10d rows, Avg time: ${stats.avgTime.toMillis}ms")
    }
    println()
  }
}

// Made with Bob

File: sdk/java/src/main/java/io/substrait/compliance/benchmark/BenchmarkExample.java
package io.substrait.compliance.benchmark;

import io.substrait.compliance.*;
import io.substrait.proto.Plan;

import java.util.*;

/**
 * Example demonstrating the benchmarking capabilities of the Substrait Compliance SDK
 */
public class BenchmarkExample {

    /**
     * Mock engine for demonstration purposes
     */
    static class MockEngine implements ComplianceEngine {
        private final String name;

        public MockEngine(String name) {
            this.name = name;
        }

        @Override
        public ComplianceResult executePlan(Plan plan, Map<String, TableData> inputData) {
            // Simulate work
            try {
                Thread.sleep(1);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
            TableData emptyTable = new TableData(
                Collections.emptyList(),
                Collections.emptyList(),
                Collections.emptyList()
            );
            return ComplianceResult.success(emptyTable, 1);
        }

        @Override
        public EngineCapabilities getCapabilities() {
            return EngineCapabilities.builder()
                    .supportsExtensions(true)
                    .build();
        }

        @Override
        public PlanValidationResult validatePlan(Plan plan) {
            return PlanValidationResult.supported();
        }

        @Override
        public EngineInfo getEngineInfo() {
            return new EngineInfo(name, "1.0.0", "0.20.0");
        }
    }

    public static void main(String[] args) throws Exception {
        System.out.println("Substrait Compliance Java SDK - Benchmark Example\n");

        MockEngine engine = new MockEngine("MockEngine");

        // Example 1: Basic operation benchmark
        System.out.println("=== Example 1: Basic Operation Benchmark ===");
        BenchmarkConfig config = BenchmarkConfig.builder()
                .warmupRuns(5)
                .measurementRuns(50)
                .verbose(true)
                .build();

        BenchmarkRunner runner = new BenchmarkRunner(engine, config);

        List<BenchmarkOperation> operations = Arrays.asList(
                BenchmarkOperation.of("simple_addition", () -> {
                    int result = 1 + 1;
                }),
                BenchmarkOperation.of("string_concat", () -> {
                    String result = "hello" + "world";
                })
        );

        BenchmarkResult result = runner.runBenchmark("basic_operations", operations);
        System.out.println(result.summary());

        // Example 2: Plan execution benchmark
        System.out.println("\n=== Example 2: Plan Execution Benchmark ===");
        Plan mockPlan = Plan.newBuilder().build();

        BenchmarkStats stats = BenchmarkRunner.quickBenchmark(
                engine,
                "execute_plan",
                () -> {
                    engine.executePlan(mockPlan, Collections.emptyMap());
                    return null;
                },
                100
        );

        System.out.println(stats.summary());

        // Example 3: Export to CSV
        System.out.println("\n=== Example 3: Export Results to CSV ===");
        String csvOutput = result.toCSV();
        System.out.println("CSV Output:");
        System.out.println(csvOutput);

        // Example 4: Comparison benchmark
        System.out.println("\n=== Example 4: Engine Comparison ===");
        MockEngine fastEngine = new MockEngine("FastEngine");
        MockEngine slowEngine = new MockEngine("SlowEngine");

        BenchmarkConfig comparisonConfig = BenchmarkConfig.builder()
                .warmupRuns(3)
                .measurementRuns(30)
                .verbose(false)
                .build();

        List<BenchmarkOperation> comparisonOps = Collections.singletonList(
                BenchmarkOperation.of("computation", () -> {
                    int sum = 0;
                    for (int i = 0; i < 1000; i++) {
                        sum += i;
                    }
                })
        );

        BenchmarkRunner fastRunner = new BenchmarkRunner(fastEngine, comparisonConfig);
        BenchmarkResult fastResult = fastRunner.runBenchmark("comparison", comparisonOps);

        BenchmarkRunner slowRunner = new BenchmarkRunner(slowEngine, comparisonConfig);
        BenchmarkResult slowResult = slowRunner.runBenchmark("comparison", comparisonOps);

        System.out.println("Fast Engine Results:");
        System.out.println(fastResult.summary());
        System.out.println("\nSlow Engine Results:");
        System.out.println(slowResult.summary());

        // Example 5: Scalability test
        System.out.println("\n=== Example 5: Scalability Test ===");
        int[] sizes = {10, 100, 1000};

        for (int size : sizes) {
            final int currentSize = size;
            List<BenchmarkOperation> scalabilityOps = Collections.singletonList(
                    BenchmarkOperation.of("process_" + size + "_items", () -> {
                        int sum = 0;
                        for (int i = 0; i < currentSize; i++) {
                            sum += i;
                        }
                    })
            );

            BenchmarkConfig scalabilityConfig = BenchmarkConfig.builder()
                    .warmupRuns(2)
                    .measurementRuns(20)
                    .verbose(false)
                    .build();

            BenchmarkRunner scalabilityRunner = new BenchmarkRunner(engine, scalabilityConfig);
            BenchmarkResult scalabilityResult = scalabilityRunner.runBenchmark(
                    "scalability_" + size,
                    scalabilityOps
            );

            System.out.println("\nSize: " + size + " items");
            for (BenchmarkStats stat : scalabilityResult.getStats()) {
                System.out.printf("  Avg: %s, Throughput: %.2f ops/sec%n",
                        formatDuration(stat.getAvgTime()),
                        stat.getThroughput());
            }
        }

        // Example 6: Parallel benchmark
        System.out.println("\n=== Example 6: Parallel Benchmark ===");
        BenchmarkConfig parallelConfig = BenchmarkConfig.builder()
                .warmupRuns(3)
                .measurementRuns(20)
                .parallelism(4)
                .verbose(false)
                .build();

        List<BenchmarkOperation> parallelOps = Arrays.asList(
                BenchmarkOperation.of("operation_1", () -> {
                    int sum = 0;
                    for (int i = 0; i < 100; i++) sum += i;
                }),
                BenchmarkOperation.of("operation_2", () -> {
                    String s = "";
                    for (int i = 0; i < 10; i++) s += "x";
                }),
                BenchmarkOperation.of("operation_3", () -> {
                    List<Integer> list = new ArrayList<>();
                    for (int i = 0; i < 50; i++) list.add(i);
                }),
                BenchmarkOperation.of("operation_4", () -> {
                    Map<String, Integer> map = new HashMap<>();
                    for (int i = 0; i < 20; i++) map.put("key" + i, i);
                })
        );

        BenchmarkRunner parallelRunner = new BenchmarkRunner(engine, parallelConfig);
        BenchmarkResult parallelResult = parallelRunner.runParallelBenchmark(
                "parallel_operations",
                parallelOps,
                4
        );

        System.out.println(parallelResult.summary());

        System.out.println("\n=== Benchmark Complete ===");
    }

    private static String formatDuration(java.time.Duration duration) {
        long nanos = duration.toNanos();
        if (nanos < 1000) {
            return nanos + "ns";
        } else if (nanos < 1_000_000) {
            return String.format("%.2fμs", nanos / 1000.0);
        } else if (nanos < 1_000_000_000) {
            return String.format("%.2fms", nanos / 1_000_000.0);
        } else {
            return String.format("%.2fs", nanos / 1_000_000_000.0);
        }
    }
}

